File: fsharp/yacht/Yacht.fs
module Yacht

type Category = 
    | Ones
    | Twos
    | Threes
    | Fours
    | Fives
    | Sixes
    | FullHouse
    | FourOfAKind
    | LittleStraight
    | BigStraight
    | Choice
    | Yacht

type Die =
    | One = 1
    | Two = 2
    | Three = 3
    | Four = 4 
    | Five = 5
    | Six = 6

let yachtScore dice =
    let firstDice = List.head dice
    let isYachtScore = List.tail dice |> List.forall ((=) firstDice);

    if isYachtScore then 50 else 0

let numberCategoryScore category dice = dice |> List.filter((=) category) |> List.length |> (*) (int category)

let fullHouseScore dice =
    let secondDice = List.last dice
    if List.distinct dice |> List.length <> 2 then
        0
    elif snd dice.Head > 3 then
        0
    else
        fst dice.Head * snd dice.Head + fst secondDice * snd secondDice

let fourOfAKindScore dice  =
    if List.distinct dice |> List.length > 2 then
        0
    elif snd dice.Head < 4 then
        0
    else
        fst dice.Head * 4


let straightScore dice minDice maxDice =
    if List.length dice = List.length (List.distinct dice) && List.min dice = minDice && List.max dice = maxDice then
        30
    else 0

let choiceScore (dice) = dice |> List.map (fun d -> int d) |> List.sum

let score category dice =
    if List.length dice <> 5 then
       failwith "Incorrect number of dice roles"
    else
        let diceTotals = dice |> List.countBy int |> List.sortByDescending snd

        match category with
        | Ones -> numberCategoryScore Die.One dice
        | Twos -> numberCategoryScore Die.Two dice
        | Threes -> numberCategoryScore Die.Three dice
        | Fours -> numberCategoryScore Die.Four dice
        | Fives -> numberCategoryScore Die.Five dice
        | Sixes -> numberCategoryScore Die.Six dice
        | FullHouse -> fullHouseScore diceTotals
        | FourOfAKind -> fourOfAKindScore diceTotals
        | LittleStraight -> straightScore dice Die.One Die.Five
        | BigStraight -> straightScore dice Die.Two Die.Six
        | Yacht -> yachtScore dice
        | Choice -> choiceScore dice

File: csharp/yacht/Yacht.cs
using System;
using System.Linq;

public enum YachtCategory
{
    Ones = 1,
    Twos = 2,
    Threes = 3,
    Fours = 4,
    Fives = 5,
    Sixes = 6,
    FullHouse = 7,
    FourOfAKind = 8,
    LittleStraight = 9,
    BigStraight = 10,
    Choice = 11,
    Yacht = 12,
}

public static class YachtGame
{ 
    public static int Score(int[] dice, YachtCategory category)
    {
        var result = category switch
        {
            YachtCategory.Yacht => YachtScore(dice),
            YachtCategory.Ones or YachtCategory.Twos or YachtCategory.Threes or YachtCategory.Fours or YachtCategory.Fives or YachtCategory.Sixes  => NumberScore(dice, category),
            YachtCategory.FullHouse => FullHouseScore(dice),
            _ => 0
        };

        return result;
    }

    private static int YachtScore(int[] dice) => dice.Distinct().Count() == 1 ? 50 : 0;

    private static int NumberScore(int[] dice, YachtCategory category) => dice.Where(diceNumber => diceNumber == (int)category).Sum();

    private static int FullHouseScore(int[] dice)
    {
        var diceNumberVariations = dice.Distinct();

        if (diceNumberVariations.Count() != 2) return 0;

        var firstDistictNumberScore = diceNumberVariations.First();
        var secondDistinctNumberScore = diceNumberVariations.Last();
        var firstDistinctNumberScoreCount = dice.Count(diceNumber => diceNumber == firstDistictNumberScore);
        var secondDistinctNumberScoreCount = dice.Count(diceNumber => diceNumber == secondDistinctNumberScore);

        if ((firstDistinctNumberScoreCount == 2 && secondDistinctNumberScoreCount == 3) || (firstDistinctNumberScoreCount == 3 && secondDistinctNumberScoreCount == 5))
            return firstDistictNumberScore * firstDistinctNumberScoreCount + secondDistinctNumberScore * secondDistinctNumberScoreCount;
        else
            return 0;
    }
}

